File: server/utils/crypto.js
import cryptoForge from 'node-forge'
import crypto from 'crypto';

/**
 * Generates a cryptographically secure random 256-bit token
 * @returns {string} A 64-character hexadecimal string representing the 256-bit token
 */
export function generate256BitToken() {
    const randomBytes = crypto.randomBytes(32);
    return randomBytes.toString('hex');
}

export function createDownloadLinks(encryptedMediaUrl) {
    if (!encryptedMediaUrl) return [];

    const qualities = [
        { id: '_12', bitrate: '12kbps' },
        { id: '_48', bitrate: '48kbps' },
        { id: '_96', bitrate: '96kbps' },
        { id: '_160', bitrate: '160kbps' },
        { id: '_320', bitrate: '320kbps' }
    ];

    const key = '38346591';
    const iv = '00000000';

    const encrypted = cryptoForge.util.decode64(encryptedMediaUrl);
    const decipher = cryptoForge.cipher.createDecipher('DES-ECB', cryptoForge.util.createBuffer(key));
    decipher.start({ iv: cryptoForge.util.createBuffer(iv) });
    decipher.update(cryptoForge.util.createBuffer(encrypted));
    decipher.finish();

    const decryptedLink = decipher.output.getBytes();

    return qualities.map((quality) => ({
        quality: quality.bitrate,
        url: decryptedLink.replace('_96', quality.id)
    }));
}

File: app/core/crypto.py
import secrets
import base64
from Crypto.Cipher import DES
from Crypto.Util.Padding import unpad

def generate_256bit_token() -> str:
    random_bytes = secrets.token_bytes(32)
    return random_bytes.hex()

def create_download_links(encrypted_media_url: str):
    if not encrypted_media_url:
        return []
    
    qualities = [
        {'id': '_12', 'bitrate': '12kbps'},
        {'id': '_48', 'bitrate': '48kbps'},
        {'id': '_96', 'bitrate': '96kbps'},
        {'id': '_160', 'bitrate': '160kbps'},
        {'id': '_320', 'bitrate': '320kbps'}
    ]
    
    key = b'38346591'
    
    try:
        encrypted_data = base64.b64decode(encrypted_media_url)
        cipher = DES.new(key, DES.MODE_ECB)
        decrypted = cipher.decrypt(encrypted_data)
        
        try:
            decrypted_link = unpad(decrypted, DES.block_size).decode('utf-8')
        except:
            decrypted_link = decrypted.decode('utf-8', errors='ignore').rstrip('\x00')
        
        return [
            {
                'quality': quality['bitrate'],
                'url': decrypted_link.replace('_96', quality['id'])
            }
            for quality in qualities
        ]
    except Exception as e:
        return []
